### DISCUSSION TITLE
How to loop choices using one button?

### DISCUSSION THREAD

**Discussion by Makima-sama** [forntoh/LcdMenu#147] [Q&A] [answered]
I'm making an LCD menu with only 2 choices which is True or False and the will only be composed of 2 buttons which is "Next" and "Enter".
My question is, does this library have a function to cycle on both choices with only using 1 button?
![Question](https://github.com/forntoh/LcdMenu/assets/143432156/928d3ef5-9056-4489-8f8b-a8578f036038)

Example Code:
// ../../examples/Basic/Basic.ino#L35-L68
// Initialize the main menu items
MAIN_MENU(
    ITEM_BASIC("TRUE"),
    ITEM_BASIC("FALSE"),
);
// Construct the LcdMenu
LcdMenu menu(LCD_ROWS, LCD_COLS);

void setup() {
    Serial.begin(9600);
    // Initialize LcdMenu with the menu items
    menu.setupLcdWithMenu(0x27, mainMenu);
}

void loop() {
    if (!Serial.available()) return;
    char command = Serial.read();
    else if (command == DOWN)
        menu.down();
    else if (command == ENTER)
        menu.enter();

}

**Comment by Amha-Endrayes** [comment, marked as answer]
Not sure if there are any built in functions with in the LCD library but you can add a custom code to your program to read if a button has been pressed once or twice to give more functionality to a single button.
i.e: if a button is pressed once move UP else if a button is pressed twice move DOWN.

There is a library to simplfy this for you [here](https://github.com/mathertel/OneButton) and [this](https://www.hackster.io/najad/single-switch-multiple-functions-20147e) might also give you some ideas.

You can implement the functionality like this:
```
// You will need this Library: https://github.com/mathertel/OneButton
#include <OneButton.h>

//===================================================================
//==================== Add This Code on Top =========================
//===================================================================
// attach the button you want to use to go up/down here:
#define BUTTON_PIN 4 

/**
 * Initialize a new OneButton instance for a button
 * connected to digital pin 4 and GND, which is active low
 * and uses the internal pull-up resistor.
 * i.e no need for an external resistor
 */

OneButton button = OneButton(
  BUTTON_PIN,  // Input pin for the button
  true,        // Button is active LOW
  true         // Enable internal pull-up resistor
);

//===================================================================

// Initialize the main menu items
MAIN_MENU(
ITEM_BASIC("TRUE"),
ITEM_BASIC("FALSE"),
);
// Construct the LcdMenu
LcdMenu menu(LCD_ROWS, LCD_COLS);

void setup() {
Serial.begin(9600);
// Initialize LcdMenu with the menu items
menu.setupLcdWithMenu(0x27, mainMenu);

//=========================================
  button.attachClicked(singleClick);
  button.attachDoubleClick(doubleClick);
//=========================================

}

void loop() {
  // keep watching the push button:
  // This is needed by the OneButton Lib
  button.tick();

  // Add the rest of your code here

  delay(10);


}

//===================================================================
//====================== ADD THIS FUNCTIONS  ========================
//===================================================================

void singleClick()
{
    menu.up();
} 


void doubleClick()
{
    menu.down();
} 

//=====================================================================
```

I notice you are using ITEM_BASIC, The menu library also have a list functionality called ITEM_STRING_LIST and that might also help you.

**Reply by forntoh** [reply]
Yep, that would work.

Or he can just check if you're on the last item in the menu while moving down and execute the up action instead of down, and vis versa.

**Reply by Amha-Endrayes** [reply]
Great idea! 
That is definitely way better

**Reply by LilTeo14** [reply]
@forntoh I was looking for some function in library that allows you to check the actual item but i cant found it, i was expecting for some function like menu.actualposition(). Thanks

**Reply by LilTeo14** [reply]
I found menu.getCursorPosition() return the cursor position, i hope it helps somebody else looking for that.

**Reply by CraigSiemens** [reply]
I also found `menu.getCursorPosition()` but I'm not seeing a function to tell what the last position of the current menu is. I was hoping to do something like `menu.getCursorPosition() == menu.getLastCursorPosition()`.

In my case there's many items in the menu, but I can only have a single button controlling selection. I also found there's `isAtTheStart` and `isAtTheEnd` which sound promising but they're private so can't be used.

**Reply by CraigSiemens** [reply]
This is the solution I ended up using. I copied what `isAtTheEnd` is doing for the condition.
```cpp
int cursorPosition = menu.getCursorPosition();
bool isAtEnd = menu.getItemAt(cursorPosition + 1)->getType() == MENU_ITEM_END_OF_MENU;
if (isAtEnd) {
  menu.resetMenu();
} else {
  menu.down();
}
```

**Reply by forntoh** [reply]
@CraigSiemens Made them public #172 🙂
